File: Sources/ClipStackCore/HistoryStore.swift
import Foundation

/// Persistent clipboard history.
///
/// Storage layout under `directory`:
///   history.json   — JSON array of `ClipItem` (newest first)
///   images/*.png   — image payloads referenced by `ClipItem.imageFileName`
///
/// NOT thread-safe by design: the app only touches it from the main thread
/// (pasteboard polling timer, menu actions and the switcher panel all run
/// on the main run loop). Tests are single-threaded too.
public final class HistoryStore {
    public private(set) var items: [ClipItem] = []
    public let maxItems: Int
    public let directory: URL
    /// Called after every mutation, on the caller's thread. Used by the UI.
    public var onChange: (() -> Void)?

    public var imagesDirectory: URL { directory.appendingPathComponent("images", isDirectory: true) }
    private var historyFile: URL { directory.appendingPathComponent("history.json") }

    public init(directory: URL, maxItems: Int = 300) throws {
        self.directory = directory
        self.maxItems = max(1, maxItems)
        let fm = FileManager.default
        try fm.createDirectory(at: directory, withIntermediateDirectories: true)
        try fm.createDirectory(at: imagesDirectory, withIntermediateDirectories: true)
        load()
    }

    // MARK: mutations

    /// If an entry with the same content already exists, move it to the front
    /// (refreshing its timestamp) instead of inserting a duplicate.
    @discardableResult
    public func promoteIfExists(hash: String, date: Date = Date()) -> Bool {
        guard let idx = items.firstIndex(where: { $0.contentHash == hash }) else { return false }
        var item = items.remove(at: idx)
        item.createdAt = date
        items.insert(item, at: 0)
        persistAndNotify()
        return true
    }

    /// Insert a new entry at the front, de-duplicating against existing content.
    public func add(_ item: ClipItem) {
        if let idx = items.firstIndex(where: { $0.contentHash == item.contentHash }) {
            // Same content re-copied: keep the existing entry (and its pin
            // state), just bump it to the front. Drop any freshly written
            // image file belonging to the duplicate.
            if let newName = item.imageFileName, newName != items[idx].imageFileName {
                try? FileManager.default.removeItem(at: imagesDirectory.appendingPathComponent(newName))
            }
            var existing = items.remove(at: idx)
            existing.createdAt = item.createdAt
            if let source = item.sourceApp { existing.sourceApp = source }
            items.insert(existing, at: 0)
        } else {
            items.insert(item, at: 0)
            trimIfNeeded()
        }
        persistAndNotify()
    }

    public func contains(hash: String) -> Bool {
        items.contains { $0.contentHash == hash }
    }

    public func delete(id: UUID) {
        guard let idx = items.firstIndex(where: { $0.id == id }) else { return }
        removeSideFiles(items[idx])
        items.remove(at: idx)
        persistAndNotify()
    }

    public func togglePin(id: UUID) {
        guard let idx = items.firstIndex(where: { $0.id == id }) else { return }
        items[idx].pinned.toggle()
        persistAndNotify()
    }

    /// Remove entries; pinned ones survive unless `keepPinned` is false.
    public func clear(keepPinned: Bool = true) {
        let victims = keepPinned ? items.filter { !$0.pinned } : items
        victims.forEach(removeSideFiles)
        items = keepPinned ? items.filter { $0.pinned } : []
        persistAndNotify()
    }

    // MARK: image payloads

    /// Persist PNG bytes and return the generated file name for `imageFileName`.
    public func storeImagePNG(_ data: Data) throws -> String {
        let name = UUID().uuidString + ".png"
        try data.write(to: imagesDirectory.appendingPathComponent(name), options: .atomic)
        return name
    }

    public func imageData(for item: ClipItem) -> Data? {
        guard let name = item.imageFileName else { return nil }
        return try? Data(contentsOf: imagesDirectory.appendingPathComponent(name))
    }

    // MARK: persistence

    public func save() {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        guard let data = try? encoder.encode(items) else { return }
        try? data.write(to: historyFile, options: .atomic)
    }

    private func load() {
        guard let data = try? Data(contentsOf: historyFile) else { return }
        if let decoded = try? JSONDecoder().decode([ClipItem].self, from: data) {
            items = decoded
        }
    }

    // MARK: internals

    private func trimIfNeeded() {
        while items.count > maxItems {
            // Evict the oldest unpinned entry first; if everything is pinned,
            // fall back to dropping the oldest entry outright.
            if let idx = items.lastIndex(where: { !$0.pinned }) {
                removeSideFiles(items[idx])
                items.remove(at: idx)
            } else {
                removeSideFiles(items.removeLast())
            }
        }
    }

    private func removeSideFiles(_ item: ClipItem) {
        guard let name = item.imageFileName else { return }
        try? FileManager.default.removeItem(at: imagesDirectory.appendingPathComponent(name))
    }

    private func persistAndNotify() {
        save()
        onChange?()
    }
}

File: windows/ClipStack.Win/HistoryStore.cs
using System.Text.Json;

namespace ClipStack.Win;

/// <summary>
/// Persistent clipboard history, newest first. Port of the macOS
/// ClipStackCore.HistoryStore: dedupe by content hash (promote to front),
/// pinned items survive trimming, JSON + PNG side files under %APPDATA%.
/// UI-thread only, same as the macOS version.
/// </summary>
public class HistoryStore
{
    public List<ClipItem> Items { get; } = new();
    public int MaxItems { get; }
    public string Directory { get; }
    public string ImagesDirectory => Path.Combine(Directory, "images");
    private string HistoryFile => Path.Combine(Directory, "history.json");

    public event Action? OnChange;

    private static readonly JsonSerializerOptions JsonOpts = new() { WriteIndented = false };

    public HistoryStore(string directory, int maxItems = 300)
    {
        Directory = directory;
        MaxItems = Math.Max(1, maxItems);
        System.IO.Directory.CreateDirectory(directory);
        System.IO.Directory.CreateDirectory(ImagesDirectory);
        Load();
    }

    public bool PromoteIfExists(string hash)
    {
        var idx = Items.FindIndex(i => i.ContentHash == hash);
        if (idx < 0) return false;
        var item = Items[idx];
        Items.RemoveAt(idx);
        item.CreatedAt = DateTime.Now;
        Items.Insert(0, item);
        PersistAndNotify();
        return true;
    }

    public void Add(ClipItem item)
    {
        var idx = Items.FindIndex(i => i.ContentHash == item.ContentHash);
        if (idx >= 0)
        {
            // Same content re-copied: keep existing entry (and pin state).
            if (item.ImageFileName != null && item.ImageFileName != Items[idx].ImageFileName)
                TryDelete(Path.Combine(ImagesDirectory, item.ImageFileName));
            var existing = Items[idx];
            Items.RemoveAt(idx);
            existing.CreatedAt = item.CreatedAt;
            if (item.SourceApp != null) existing.SourceApp = item.SourceApp;
            Items.Insert(0, existing);
        }
        else
        {
            Items.Insert(0, item);
            TrimIfNeeded();
        }
        PersistAndNotify();
    }

    public void Delete(Guid id)
    {
        var idx = Items.FindIndex(i => i.Id == id);
        if (idx < 0) return;
        RemoveSideFiles(Items[idx]);
        Items.RemoveAt(idx);
        PersistAndNotify();
    }

    public void TogglePin(Guid id)
    {
        var item = Items.FirstOrDefault(i => i.Id == id);
        if (item == null) return;
        item.Pinned = !item.Pinned;
        PersistAndNotify();
    }

    public void Clear(bool keepPinned = true)
    {
        foreach (var victim in Items.Where(i => !keepPinned || !i.Pinned))
            RemoveSideFiles(victim);
        Items.RemoveAll(i => !keepPinned || !i.Pinned);
        PersistAndNotify();
    }

    public string StoreImagePng(byte[] data)
    {
        var name = Guid.NewGuid().ToString("N") + ".png";
        File.WriteAllBytes(Path.Combine(ImagesDirectory, name), data);
        return name;
    }

    public byte[]? ImageData(ClipItem item)
    {
        if (item.ImageFileName == null) return null;
        var path = Path.Combine(ImagesDirectory, item.ImageFileName);
        return File.Exists(path) ? File.ReadAllBytes(path) : null;
    }

    private void TrimIfNeeded()
    {
        while (Items.Count > MaxItems)
        {
            var idx = Items.FindLastIndex(i => !i.Pinned);
            if (idx < 0) idx = Items.Count - 1; // everything pinned: drop oldest
            RemoveSideFiles(Items[idx]);
            Items.RemoveAt(idx);
        }
    }

    private void RemoveSideFiles(ClipItem item)
    {
        if (item.ImageFileName != null)
            TryDelete(Path.Combine(ImagesDirectory, item.ImageFileName));
    }

    private static void TryDelete(string path)
    {
        try { if (File.Exists(path)) File.Delete(path); } catch { /* best effort */ }
    }

    private void PersistAndNotify()
    {
        Save();
        OnChange?.Invoke();
    }

    public void Save()
    {
        try
        {
            File.WriteAllText(HistoryFile, JsonSerializer.Serialize(Items, JsonOpts));
        }
        catch { /* best effort */ }
    }

    private void Load()
    {
        try
        {
            if (!File.Exists(HistoryFile)) return;
            var loaded = JsonSerializer.Deserialize<List<ClipItem>>(File.ReadAllText(HistoryFile));
            if (loaded != null) Items.AddRange(loaded);
        }
        catch { /* corrupt history: start fresh */ }
    }
}
